### PR TITLE
move verification and sanitization of WITH clause to one place

### PR DESCRIPTION
**Pull request by agavra**
### Description 
This change just materializes and performs validation of the WITH clause in one place: `StatementProperties` so that we don't have to always access through a `Map<String, Literal>` and instead use a model. The rest makes small changes around that. 

**NOTE**: this makes it so that you cannot use `WITH(registered_topic="foo")` anymore, though this was never documented and I believe it didn't work before this change anyway

### Testing done 
- Existing and new unit tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/AbstractCreateStreamCommand.java [hunk 1]
```diff
@@ -91,8 +74,8 @@
 
     this.schema = getStreamTableSchema(statement.getElements());
 
-    if (properties.containsKey(DdlConfig.KEY_NAME_PROPERTY)) {
-      final String name = properties.get(DdlConfig.KEY_NAME_PROPERTY).toString().toUpperCase();
+    if (properties.getKey().isPresent()) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/AbstractCreateStreamCommand.java]: `KEY` defines the _key field_, so maybe:

```suggestion
    if (properties.getKeyField().isPresent()) {
```

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/AbstractCreateStreamCommand.java [hunk 2]
```diff
@@ -165,64 +138,26 @@ static void checkMetaData(
...
   ) {
-    final Literal topicNameExp = properties.get(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY);
-
-    if (topicNameExp == null) {
-      throw new KsqlException("Corresponding Kafka topic ("
-          + DdlConfig.KAFKA_TOPIC_NAME_PROPERTY + ") should be set in WITH clause.");
-    }
-
-    final String kafkaTopicName = StringUtil.cleanQuotes(topicNameExp.toString());
+    final String kafkaTopicName = StringUtil.cleanQuotes(properties.getKafkaTopic());
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/AbstractCreateStreamCommand.java]: `StringUtil.cleanQuotes` could be done inside the `getKafkaTopic` call

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java [hunk 3]
```diff
@@ -397,16 +397,7 @@ protected Node visitCreateStream(final CreateStream node, final Object context)
...
-        .getProperties()
-        .entrySet()
-        .stream()
-        .collect(Collectors.toMap(
-            Entry::getKey,
-            e -> (Literal) process(e.getValue(), context)
-        ));
-
-    return node.copyWith(rewrittenElements, rewrittenProps);
+    return node.copyWith(rewrittenElements, node.getProperties());
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java]: This is no longer allowing the literals to be re-written. Is this intentional?

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java [hunk 4]
```diff
@@ -433,16 +424,7 @@ protected Node visitCreateTable(final CreateTable node, final Object context) {
...
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java]: This is no longer allowing the literals to be re-written. Is this intentional?

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/AbstractStreamCreateStatement.java [hunk 5]
```diff
@@ -31,23 +30,38 @@
...
-  private final ImmutableMap<String, Literal> properties;
+  private final StatementProperties properties;
 
   AbstractStreamCreateStatement(
       final Optional<NodeLocation> location,
       final QualifiedName name,
       final List<TableElement> elements,
       final boolean notExists,
       final Map<String, Literal> properties
+  ) {
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/tree/AbstractStreamCreateStatement.java]: can we not just drop the old constructor?

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/CreateStream.java [hunk 6]
```diff
@@ -44,10 +44,20 @@ public CreateStream(
     super(location, name, elements, notExists, properties);
   }
 
+  private CreateStream(
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/tree/CreateStream.java]: likewise, can we not drop the old constructor?

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/CreateTable.java [hunk 7]
```diff
@@ -44,10 +44,20 @@ public CreateTable(
     super(location, name, elements, notExists, properties);
   }
 
+  private CreateTable(
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/tree/CreateTable.java]: likewise, can we not drop the old constructor?

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/StatementProperties.java [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+      final Property<String> ksqlTopic,
+      final Property<String> timestampName,
+      final Property<String> windowType,
+      final Property<String> key,
+      final Property<String> kafkaTopic,
+      final Property<Format> valueFormat,
+      final Property<String> avroSchemaName,
+      final Property<Integer> avroSchemaId,
+      final Property<String> timestampFormat) {
+    this.kafkaTopic = kafkaTopic;
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/tree/StatementProperties.java]: nit: null checks :D and remove blank lines...

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/StatementProperties.java [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+  @Override
+  public int hashCode() {
+    return Objects
+        .hash(kafkaTopic, valueFormat, key, ksqlTopic, windowType, timestampName, timestampFormat,
+            avroSchemaName, avroSchemaId);
+  }
+
+  private static final class Property<T> {
+
+    final T value;
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/tree/StatementProperties.java]: nit: private

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/StatementProperties.java [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+    final String name;
+    final Literal original;
+
+    private Property(final String name, final Literal literal , final Function<Object, T> extract) {
+      this.original = literal;
+      this.value = literal == null ? null : cleanQuotesIfString(extract.apply(literal.getValue()));
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/tree/StatementProperties.java]: `literal.getValue()` would return a string _without_ the quotes.  So the `cleanQuotesIfString` should not be required.

File: ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java [hunk 11]
```diff
@@ -498,14 +499,14 @@ public void testCreateStreamWithTopic() {
     final String
         queryStr =
         "CREATE STREAM orders (ordertime bigint, orderid varchar, itemid varchar, orderunits "
-        + "double) WITH (registered_topic = 'orders_topic' , key='ordertime');";
+        + "double) WITH (registered_topic = 'orders_topic' , key='ordertime', kafka_topic='foo', value_format='json');";
```
Comment (big-andy-coates) [inline ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java]: I thought `registered_topic` was no longer supported...

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/AbstractCreateStreamCommand.java (hunk 2 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/AbstractCreateStreamCommand.java]: ah yes! it is being done :)

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java (hunk 3 above)
Comment (agavra) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java]: yes it is - I noticed that nowhere was using this and I don't think it really makes sense to arbitrarily rewrite literals with a rewriter. If you want to rewrite the properties, I would suggest doing it manually. If you feel strongly about this I can rework it but it's not a simple task.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java (hunk 4 above)
Comment (agavra) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriter.java]: See above

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/AbstractStreamCreateStatement.java (hunk 5 above)
Comment (agavra) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/tree/AbstractStreamCreateStatement.java]: I suppose we can :)

File: ksql-parser/src/main/java/io/confluent/ksql/parser/tree/CreateStream.java (hunk 6 above)
Comment (agavra) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/tree/CreateStream.java]: They're being used all over the code for testing, and I think it's a useful utility method to have.

File: ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java (hunk 11 above)
Comment (agavra) [reply ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java]: That's what I thought as well. I just kept this in tact - if we want to remove it from syntax that's for another PR

**Comment by agavra** [comment]
> One thing to note is that the set of properties allowed is different for different statements. This mainly being C* vs CAS statements allowing different props, e.g. you should not be able to set the KEY property in CAS statements.

Yup, I noticed this as well - wasn't quite sure how to handle it but figured this is better than not this :) As far as the approaches you listed out, I'm also really unsure which one is better...

> That's a bit of a ramble, so what am I actually saying in the context of this PR? Well, probably just rename StatementProperties to CreateStatementProperties. We can look to add a CreateAsStatementProperties later to cover the other set, following a similar pattern.

That is sound.

**Comment by agavra** [comment]
Merged with 1 review since there is no functional change and this unblocks future changes

**Pull request merged**